src/engine/math_util.h

```diff
@@ -61,90 +56,118 @@ extern f32 gSineTable[];
-// #define max(a, b) MAX((a), (b)) // ((a) > (b) ? (a) : (b))
-#define CLAMP(x, low, high)  (((x) > (high)) ? (high) : (((x) < (low)) ? (low) : (x)))
-
-// from limits.h
-#define S8_MAX __SCHAR_MAX__
-#define S8_MIN (-S8_MAX - 1)
-#define U8_MAX (S8_MAX * 2 + 1)
-#define S16_MAX __SHRT_MAX__
-#define S16_MIN (-S16_MAX - 1)
-#define U16_MAX (S16_MAX * 2 + 1)
-#define S32_MAX __INT_MAX__
-#define S32_MIN (-S32_MAX - 1)
-#define U32_MAX (S32_MAX * 2U + 1U)
-#define S64_MAX __LONG_LONG_MAX__
-#define S64_MIN (-S64_MAX - 1LL)
-#define U64_MAX (S64_MAX * 2ULL + 1ULL)
-#define F32_MAX __FLT_MAX__
-#define F32_MIN __FLT_MIN__
-#define F64_MAX __DBL_MAX__
-#define F64_MIN __DBL_MIN__
-
-#define CLAMP_U8( x)        CLAMP((x),     0x0,  U8_MAX)
+
+// Various basic helper macros
+
+// Get the square of a number
+#define sqr(x) ({         \
+    __auto_type _x = (x); \
+    _x * _x; })
+
+// Get the sign of a number
+#define signum_positive(x) (((x) < 0) ? -1 : 1)
+
+// Absolute value
+#define ABS(x) ({         \
+    __auto_type _x = (x); \
+    _x > 0 ? _x : -_x; })
+#define absi ABS
+#define abss ABS
+
+// Absolute value of a float (faster than using the above macro)
+ALWAYS_INLINE f32 absf(f32 in) {
+    f32 out;
+    __asm__("abs.s %0,%1" : "=f" (out) : "f" (in));
+    return out;
+}
+
+// Get the minimum / maximum of a set of numbers
+#undef MIN
+#define MIN(a, b) ({      \
+    __auto_type _a = (a); \
+    __auto_type _b = (b); \
+    _a < _b ? _a : _b; })
+
+#undef MAX
+#define MAX(a, b) ({      \
+    __auto_type _a = (a); \
+    __auto_type _b = (b); \
+    _a > _b ? _a : _b; })
+
+#define min_3(a, b, c) MIN(MIN(a, b), c)
+
+#define max_3(a, b, c) MAX(MAX(a, b), c)
+
+#define min_3f min_3
+#define min_3i min_3
+#define min_3s min_3
+
+#define max_3f max_3
+#define max_3i max_3
+#define max_3s max_3
+
+void min_max_3f(f32 a, f32 b, f32 c, f32 *min, f32 *max);
+void min_max_3i(s32 a, s32 b, s32 c, s32 *min, s32 *max);
+void min_max_3s(s16 a, s16 b, s16 c, s16 *min, s16 *max);
+
+// From Wiseguy
+// Round a float to the nearest integer
+ALWAYS_INLINE s32 roundf(f32 in) {
+    f32 tmp;
+    s32 out;
+    __asm__("round.w.s %0,%1" : "=f" (tmp) : "f" (in ));
+    __asm__("mfc1      %0,%1" : "=r" (out) : "f" (tmp));
+    return out;
+}
+
+#define round_float roundf
+
+#define FLT_IS_NONZERO(x) (absf(x) > NEAR_ZERO)
+
+
+// Integer limits and clamping
+
+#define S8_MAX 127
+#define S8_MIN -128
+#define U8_MAX 255
+#define S16_MAX 32767
+#define S16_MIN -32768
+#define U16_MAX 65535
+
+// Clamp a value inbetween a range
+#define CLAMP(x, low, high)  MIN(MAX((x), (low)), (high))
+
+// Clamp a value to the range of a specific data type
+#define CLAMP_U8( x)        CLAMP((x),       0,  U8_MAX)
 #define CLAMP_S8( x)        CLAMP((x),  S8_MIN,  S8_MAX)
-#define CLAMP_U16(x)        CLAMP((x),     0x0, U16_MAX)
+#define CLAMP_U16(x)        CLAMP((x),       0, U16_MAX)
 #define CLAMP_S16(x)        CLAMP((x), S16_MIN, S16_MAX)
```

Comment:
I'd leave the 32-bit options in here at least (also why bother changing the previous definitions?)

Reply:
theyre pretty useless since people are not going to be using 64 bit numbers much

Reply:
Clamping maybe, but max and min I would disagree with especially for signed numbers.

Reply:
added them back